File: lightsoff/gauss.hs
-- http://en.wikipedia.org/wiki/Gaussian_elimination
{-
for k = 1 ... min(m,n):
   Find the k-th pivot:
   i_max  := argmax (i = k ... m, abs(A[i, k]))
   if A[i_max, k] = 0
     error "Matrix is singular!"
   swap rows(k, i_max)
   Do for all rows below pivot:
   for i = k + 1 ... m:
     Do for all remaining elements in current row:
     for j = k + 1 ... n:
       A[i, j]  := A[i, j] - A[k, j] * (A[i, k] / A[k, k])
     Fill lower triangular matrix with zeros:
     A[i, k]  := 0
-}

{-
gauss 1 rs = rs
gauss k ((a_kk:r):rs) = r : gauss (k-1) (eliminate c (map (splitAt 1) rs))
  eliminate a_kk a_ik (c:cs) =
-}

-- All-Integer Echelon
-- http://www.math.iupui.edu/~momran/m118/integers.html
-- http://en.wikipedia.org/wiki/Gaussian_elimination
-- http://en.literateprograms.org/Row_echelon_form_%28Haskell%29

--vsub :: (Fractional a) => [a] -> [a] -> [a]
vsub a b = zipWith (-) a b

--vscale  :: (Fractional a) => a -> [a] -> [a]
vscale c v = map (*c) v

--elementryRowOpperation :: (Fractional a) => [a] -> [a] -> [a]
elementryRowOpperation us vs = vsub vs (vscale (v/u) us)
     where u = head us
           v = head vs

--ref :: (Fractional a) => [[a]] -> [[a]]
ref (ms:[]) = [ms]
ref (ms:mss) = ms : zipWith (:) (map head nss) (ref (map tail nss))
      where nss = map (elementryRowOpperation ms) mss

m = [
    [2.0,1.0,3.0,5.0],
    [0.0,2.0,-1.0,5.0],
    [2.0,1.0,2.0,6.0]
  ]

-- example from http://en.wikipedia.org/wiki/Gaussian_elimination
m2 = [
    [2.0, 1.0, -1.0, 8],
    [-3.0, -1.0, 2.0, -11.0],
    [-2.0, 1.0, 2.0, -3.0]
  ]

-- [[2.0,1.0,-1.0,8.0],[0.0,0.5,0.5,1.0],[0.0,0.0,-1.0,1.0]]


-- https://haskellicious.wordpress.com/2012/11/26/the-gauss-algorithm-in-haskell/

type Number = Double
type Vector = [Number]
type Row    = [Number]
type Matrix = [Row]

mapMatrix :: Matrix -> Vector -> Vector
mapMatrix rows v = [sum (zipWith (*) row v) | row <- rows]

exampleA = [[1,1,0], [0,1,1], [1,0,1]] :: Matrix
exampleb = [2,3,4] :: Vector
--mapMatrix exampleA x == exampleb

gauss :: Matrix -> Vector -> Vector
gauss a b = x
   where
     b' = map (\y -> [y]) b
     a' = zipWith (++) a b'    -- combine with right-hand side
     x  = resubstitute $ triangular a'

triangular :: Matrix -> Matrix
triangular [] = []
triangular m  = row:(triangular rows')
    where
    (row:rows) = rotatePivot m    -- see discussion below
    rows' = map f rows
    f bs
        | (head bs) == 0 = drop 1 bs
        | otherwise      = drop 1 $ zipWith (-) (map (*c) bs) row
        where
        c = (head row)/(head bs)    -- suitable multiple

rotatePivot :: Matrix -> Matrix
rotatePivot (row:rows)
    | (head row) /= 0 = (row:rows)
    | otherwise       = rotatePivot (rows ++ [row])

resubstitute :: Matrix -> Vector
resubstitute = reverse . resubstitute' . reverse . map reverse

resubstitute' :: Matrix -> Vector
resubstitute' [] = []
resubstitute' (row:rows) = x:(resubstitute' rows')
    where
      x     = (head row)/(last row)
      rows' = map substituteUnknown rows
      substituteUnknown (a1:(a2:as')) = ((a1-x*a2):as')

--let x = gauss exampleA exampleb
--let yyyy = mapMatrix exampleA x == exampleb

File: lightsoff/gauss.js
function print(M, msg) {
  console.log("======" + msg + "=========")
  for(var k=0; k<M.length; ++k) {
    console.log(M[k]);
  }
  console.log("==========================")
}

function diagonalize(M) {
  var m = M.length;
  var n = M[0].length;
  for(var k=0; k<Math.min(m,n); ++k) {
    // Find the k-th pivot
    i_max = findPivot(M, k);
    if (A[i_max] & (1 << k) == 0)
      throw "matrix is singular";
    swap_rows(M, k, i_max);
    // Do for all rows below pivot
    for(var i=k+1; i<m; ++i) {
      // Do for all remaining elements in current row:
      var c = A[i][k] / A[k][k];
      for(var j=k+1; j<n; ++j) {
        A[i][j] = A[i][j] - A[k][j] * c;
      }
      // Fill lower triangular matrix with zeros
      A[i][k] = 0;
    }
  }
}

function findPivot(M, k) {
  var i_max = k;
  for(var i=k+1; i<M.length; ++i) {
    if (Math.abs(M[i][k]) > Math.abs(M[i_max][k])) {
      i_max = i;
    }
  }
  return i_max;
}

function swap_rows(M, i_max, k) {
  if (i_max != k) {
    var temp = A[i_max];
    A[i_max] = A[k];
    A[k] = temp;
  }
}

function makeM(A, b) {
  for(var i=0; i<A.length; ++i) {
    A[i].push(b[i]);
  }
}

function substitute(M) {
  var m = M.length;
  for(var i=m-1; i>=0; --i) {
    var x = M[i][m] / M[i][i];
    for(var j=i-1; j>=0; --j) {
      M[j][m] -= x * M[j][i];
      M[j][i] = 0;
    }
    M[i][m] = x;
    M[i][i] = 1;
  }
}

function extractX(M) {
  var x = [];
  var m = A.length;
  var n = A[0].length;
  for(var i=0; i<m; ++i){
    x.push(A[i][n-1]);
  }
  return x;
}

function solve(A, b) {
  //print(A, "A");
  makeM(A,b);
  //print(A, "M");
  diagonalize(A);
  //print(A, "diag");
  substitute(A);
  //print(A, "subst");
  var x = extractX(A);
  //print(x, "x");
  return x;
}

// sample from: http://mathworld.wolfram.com/GaussianElimination.html

A = [
  [9,3,4],
  [4,3,4],
  [1,1,1]
]

b = [7,8,3]

print(A, " A ");
print(b, " b ");

var x = solve(A, b);

print(x, " x ");
